File: aidl/hwcontrol/java_extension/src/com/harshit/nabuextensions/stylus/StylusSettingsFragment.java
/*
 * Copyright (C) 2018,2020 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.harshit.nabuextensions.stylus;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.SystemProperties;

import androidx.preference.Preference;
import androidx.preference.Preference.OnPreferenceChangeListener;
import androidx.preference.PreferenceFragment;
import androidx.preference.SwitchPreferenceCompat;

import com.harshit.nabuextensions.HwStateManager;
import custom.hardware.hwcontrol.HwType;

import com.harshit.nabuextensions.R;

public class StylusSettingsFragment extends PreferenceFragment implements
        OnPreferenceChangeListener {

    private static final String STYLUS_KEY = "stylus_switch_key";
    private static final String STYLUS_GEN_KEY = "stylus_gen_switch_key";
    public static final String SHARED_STYLUS = "shared_stylus";
    public static final String SHARED_STYLUS_GEN = "shared_stylus_gen";
    public static final String SHARED_STYLUS_GEN_PROP = "persist.mi_pen.gen";

    private SwitchPreferenceCompat mStylusPreference;
    private SwitchPreferenceCompat mStylusGenPreference;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.stylus_settings);

        mStylusPreference = (SwitchPreferenceCompat) findPreference(STYLUS_KEY);
        mStylusPreference.setEnabled(true);
        mStylusPreference.setOnPreferenceChangeListener(this);

        mStylusGenPreference = (SwitchPreferenceCompat) findPreference(STYLUS_GEN_KEY);
        mStylusGenPreference.setEnabled(true);
        mStylusGenPreference.setOnPreferenceChangeListener(this);
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue) {
        if (STYLUS_KEY.equals(preference.getKey())) {
            enableStylus((Boolean) newValue ? 1 : 0);
        }

        if (STYLUS_GEN_KEY.equals(preference.getKey())) {
            setStylusGen((Boolean) newValue ? 1 : 0);
        }

        return true;
    }

    private void enableStylus(int status) {
        try {
            HwStateManager.HwState(HwType.STYLUS, status);
            SharedPreferences preferences = getActivity().getSharedPreferences(SHARED_STYLUS, Context.MODE_PRIVATE);
            SharedPreferences.Editor editor = preferences.edit();
            editor.putInt(SHARED_STYLUS, status);
            editor.commit();
        } catch (Exception e) {
        }
    }

    private void setStylusGen(int status) {
        String gen = status == 1 ? "2" : "1";
        try {
            SystemProperties.set(SHARED_STYLUS_GEN_PROP, gen);
            SharedPreferences preferences = getActivity().getSharedPreferences(SHARED_STYLUS_GEN, Context.MODE_PRIVATE);
            SharedPreferences.Editor editor = preferences.edit();
            editor.putString(SHARED_STYLUS_GEN, gen);
            editor.commit();
        } catch (Exception e) {
        }
    }
}

File: aidl/hwcontrol/java_extension/src/com/harshit/nabuextensions/stylus/StylusSettingsFragment.kt
/*
 * Copyright (C) 2018,2020 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.harshit.nabuextensions.stylus

import android.os.SystemProperties
import androidx.preference.SwitchPreferenceCompat
import com.harshit.nabuextensions.BaseHardwareFragment
import com.harshit.nabuextensions.HwStateManager
import com.harshit.nabuextensions.PreferenceManager
import com.harshit.nabuextensions.R
import custom.hardware.hwcontrol.HwType

class StylusSettingsFragment : BaseHardwareFragment() {
    
    private val stylusGenKey: String
        get() = findPreference<SwitchPreferenceCompat>(STYLUS_GEN_KEY)?.key ?: ""
    
    override fun getPreferenceResource(): Int = R.xml.stylus_settings
    
    override fun getSwitchKey(): String = STYLUS_KEY
    
    override fun getAdditionalSwitches(): Map<String, SwitchPreferenceCompat> {
        val genSwitch = findPreference<SwitchPreferenceCompat>(STYLUS_GEN_KEY)
        return if (genSwitch != null) {
            mapOf(STYLUS_GEN_KEY to genSwitch)
        } else {
            emptyMap()
        }
    }
    
    override fun onSwitchChanged(key: String, newValue: Boolean) {
        when (key) {
            STYLUS_KEY -> enableStylus(newValue.toInt())
            STYLUS_GEN_KEY -> setStylusGeneration(newValue)
        }
    }
    
    private fun enableStylus(status: Int) {
        HwStateManager.setHwState(HwType.STYLUS, status)
        PreferenceManager.setStylusEnabled(requireContext(), status)
    }
    
    private fun setStylusGeneration(gen2: Boolean) {
        val generation = if (gen2) "2" else "1"
        SystemProperties.set(PreferenceManager.KEY_STYLUS_GEN_PROP, generation)
        PreferenceManager.setStylusGeneration(requireContext(), generation)
    }
    
    companion object {
        private const val STYLUS_KEY = "stylus_switch_key"
        private const val STYLUS_GEN_KEY = "stylus_gen_switch_key"
    }
}

